custom ops: don't error if autograd input is a tensor subclass

bdhirsh opened the pull request:
This is needed to allow the custom ops in our custom op autograd tests to accept `FunctionalTensor` arguments as inputs that we compute gradients for. Previously, custom ops would raise an error if you tried to pass in a tensor subclass when using autograd.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105236
* #104974
* #104483
* #106406
* #109662
* #109023
* #108654
* #109632
* #108656
* __->__ #109248
* #109024

bdhirsh committed:
custom ops: don't error if autograd input is a tensor subclass

[ghstack-poisoned]

bdhirsh committed:
Update on "custom ops: don't error if autograd input is a tensor subclass"




[ghstack-poisoned]

bdhirsh committed:
Update on "custom ops: don't error if autograd input is a tensor subclass"

This is needed to allow the custom ops in our custom op autograd tests to accept `FunctionalTensor` arguments as inputs that we compute gradients for. Previously, custom ops would raise an error if you tried to pass in a tensor subclass when using autograd.




[ghstack-poisoned]

zou3519 commented on torch/_custom_op/autograd.py:
```diff
@@ -226,10 +221,6 @@ def error(what):
             error(f"got object of type {type(grad)} as the gradient for input "
                   f"'{name}', "
                   f"but expected the gradient to be either None or a Tensor")
-        if arg_info != torch.Tensor:
-            error(f"got a Tensor as the gradient for input '{name}' but "
-                  f"expected None as the gradient because input '{name}' "
-                  f"was not a Tensor (it was type {arg_info}).")
```
What this code is doing is:
- earlier, we recorded the types of all the args to the operator. the types are listed in arg_info
- during the backward pass, this check is checking that if the gradient is a Tensor, then the arg we are computing the gradient for is also a Tensor

It's not really clear to me what exactly the problem you're running into is, but it is something along the lines of "this check for if the original arg was a Tensor is too restrictive". To fix the problem, we need to relax the check.

To relax the check, we should instead change arg_info to be a boolean if the input is a {Tensor, Tensor subclass} instead of storing the type of the input.

ezyang commented:
deferring to @zou3519 

a test would help!

bdhirsh committed:
Update on "custom ops: don't error if autograd input is a tensor subclass"

This is needed to allow the custom ops in our custom op autograd tests to accept `FunctionalTensor` arguments as inputs that we compute gradients for. Previously, custom ops would raise an error if you tried to pass in a tensor subclass when using autograd.




[ghstack-poisoned]

bdhirsh committed:
Update on "custom ops: don't error if autograd input is a tensor subclass"

This is needed to allow the custom ops in our custom op autograd tests to accept `FunctionalTensor` arguments as inputs that we compute gradients for. Previously, custom ops would raise an error if you tried to pass in a tensor subclass when using autograd.




[ghstack-poisoned]

bdhirsh committed:
Update on "custom ops: don't error if autograd input is a tensor subclass"

This is needed to allow the custom ops in our custom op autograd tests to accept `FunctionalTensor` arguments as inputs that we compute gradients for. Previously, custom ops would raise an error if you tried to pass in a tensor subclass when using autograd.




[ghstack-poisoned]

zou3519 commented on torch/_custom_op/autograd.py:
```diff
@@ -226,11 +226,12 @@ def error(what):
                   f"'{name}', "
                   f"but expected the gradient to be either None or a Tensor")
-        if arg_info != torch.Tensor:
+        if not issubclass(arg_info, torch.Tensor):
```
Can you add a test?

bdhirsh committed:
Update on "custom ops: don't error if autograd input is a tensor subclass"

This is needed to allow the custom ops in our custom op autograd tests to accept `FunctionalTensor` arguments as inputs that we compute gradients for. Previously, custom ops would raise an error if you tried to pass in a tensor subclass when using autograd.




[ghstack-poisoned]

bdhirsh committed:
Update on "custom ops: don't error if autograd input is a tensor subclass"

This is needed to allow the custom ops in our custom op autograd tests to accept `FunctionalTensor` arguments as inputs that we compute gradients for. Previously, custom ops would raise an error if you tried to pass in a tensor subclass when using autograd.




[ghstack-poisoned]

bdhirsh committed:
Update on "custom ops: don't error if autograd input is a tensor subclass"

This is needed to allow the custom ops in our custom op autograd tests to accept `FunctionalTensor` arguments as inputs that we compute gradients for. Previously, custom ops would raise an error if you tried to pass in a tensor subclass when using autograd.




[ghstack-poisoned]
